dlgr/griduniverse/experiment.py

```diff
@@ -731,77 +738,135 @@ def instructions(self):
         if not position:
-            position = self._random_empty_position()
+            position = self._random_empty_position(item_id)
 
-        self.food_locations[tuple(position)] = Food(
-            id=(len(self.food_locations) + len(self.food_consumed)),
+        item_props = self.item_config[item_id]
+        new_item = Item(
+            id=(len(self.item_locations) + len(self.items_consumed)),
             position=position,
-            maturation_speed=self.food_maturation_speed,
+            item_config=item_props,
         )
-        self.food_updated = True
-        self.log_event({
-            'type': 'spawn_food',
-            'position': position,
-        })
-
-    def food_changed(self, last_food):
-        locations = self.food_locations
-        if len(last_food) != len(locations):
+        self.item_locations[tuple(position)] = new_item
+        self.items_updated = True
+        logger.warning(f"Spawning new item: {new_item}")
+        self.log_event(
+            {
+                "type": "spawn item",
+                "position": position,
+            }
+        )
+
+    def locations_of_item_of_type(self, item_id):
+        """Return"""
+        return {k: v for k, v in self.item_locations.items() if v.item_id == item_id}
+
+    def items_changed(self, last_items):
+        locations = self.item_locations
+        if len(last_items) != len(locations):
             return True
-        for food in last_food:
-            position = tuple(food['position'])
+        for item in last_items:
+            position = tuple(item["position"])
             if position not in locations:
                 return True
             found = locations[position]
-            if found.id != food['id'] or found.maturity != food['maturity']:
+            if found.id != item["id"] or found.maturity != item["maturity"]:
                 return True
         return False
 
+    def replenish_items(self):
+        for item_type in self.item_config.values():
+            # Alternate positive and negative growth rates
+            seasonal_growth = item_type["seasonal_growth_rate"] ** (
+                -1 if self.round % 2 else 1
+            )
+            logger.warning(
+                f"item_type: {item_type['name']}, seasonal_growth: {seasonal_growth}"
+            )
+            # Compute how many items of this type we should have on the grid,
+            # ensuring it's not less than zero.
+            item_type["item_count"] = max(
+                min(
+                    item_type["item_count"] * item_type["spawn_rate"] * seasonal_growth,
+                    self.rows * self.columns,
+                ),
+                0,
+            )
+            logger.warning(
+                f"item_type: {item_type['name']}, target count: {item_type['item_count']}"
+            )
+
+            # Only items of the same type.
+            item_locations = self.locations_of_item_of_type(item_type["item_id"])
```

Comment:
Since we loop over all `n` items for every `m` item types, we're `m*n` here. It looks like this is only used in this context. So maybe we'd be better off generating a `locations_by_type` map with a single loop over the `n` items before we loop over types?

Reply:
Fixed